ctapipe/io/metadata.py

```diff
@@ -0,0 +1,188 @@
+"""
+Management of CTA Reference Metadata, as defined in the CTA Top-Level Data Model
+document, version 1A. This information is required to be attached to the header of any
+files generated.
+
+The class Reference collects all required reference metadata, and can be turned into a
+flat dictionary.  The user should try to fill out all fields, or use a helper to fill
+them (as in `Activity.from_provenance()`)
+
+.. code-block:: python
+
+    ref = Reference(
+        contact=Contact(name="Some User", email="user@me.com"),
+        product=Product(format='hdf5', ...),
+        process=Process(...),
+        activity=Activity(...),
+        instrument = Instrument(...)
+    )
+
+    some_astropy_table.meta = ref.to_dict()
+    some_astropy_table.write("output.ecsv")
+
+"""
+import uuid
+from collections import OrderedDict
+
+from astropy.time import Time
+from traitlets import (
+    Enum,
+    Unicode,
+    Int,
+    HasTraits,
+    default,
+    Instance,
+    validate,
+)
+
+__all__ = [
+    "Reference",
+    "Contact",
+    "Process",
+    "Product",
+    "Activity",
+    "Instrument",
+    "write_to_hdf5",
+]
+
+
+class Contact(HasTraits):
+    """ Contact information """
+
+    name = Unicode("unknown")
+    email = Unicode("unknown")
+    organization = Unicode("unknown")
+
+
+class Product(HasTraits):
+    """Data product information"""
+
+    description = Unicode("unknown")
+    creation_time = Unicode()
+    _id = Unicode(help="leave unspecified to automatically generate a UUID")
+    data_category = Enum(["S", "A", "B", "C", "Other"], "Other")
+    data_level = Enum(
+        ["R0", "R1", "DL0", "DL1", "DL2", "DL3", "DL4", "DL5", "DL6", "Other"], "Other"
+    )
+    data_association = Enum(["Subarray", "Telescope", "Target", "Other"], "Other")
+    data_model_name = Unicode("unknown")
+    data_model_version = Unicode("unknown")
+    data_model_url = Unicode("unknown")
+    format = Unicode()
+
+    @default("creation_time")
+    def default_time(self):
+        return Time.now().iso
+
+    @validate("creation_time")
+    def valid_time(selfs, proposal):
+        thetime = Time(proposal["value"])
+        return thetime.iso
+
+    @default("_id")
+    def default_product_id(self):
+        return str(uuid.uuid4())
+
+
+class Process(HasTraits):
+    """ Process (top-level workflow) information """
+
+    _type = Enum(["Observation", "Simulation", "Other"], "Other")
+    subtype = Unicode("")
+    _id = Int()
+
+
+class Activity(HasTraits):
+    """ Activity (tool) information """
+
+    @classmethod
+    def from_provenance(cls, activity: "ActivityProvenance"):
+        return Activity(
+            name=activity["activity_name"],
+            _type="software",
+            _id=activity["activity_uuid"],
+            start=activity["start"]["time_utc"],
+            software_name="ctapipe",
+            software_version=activity["system"]["ctapipe_version"],
+        )
+
+    name = Unicode()
+    _type = Unicode("software")
```

Comment:
Are these private? Or is the underscore just to avoid overriding python builtins? If the latter, I would put the underscore at the back to not make people think these are somehow private.

Reply:
Also overriding builtins in a class namespace is not really an issue. It will only shadow the builtin in the class block, not even in the methods.

Reply:
ok will do (underscore after).  Yes, it was to avoid pylint from complaining about the builtin overrides, but it's true I could just ignore them, but it could lead to some confusion.